File: javascript/packages/client/test/contract.test.ts
import { describe, test, expect, beforeEach } from "vitest"

import { SlotIndex } from "../src/slot-index"
import type { DependencyMap } from "../src/state"
import type { Item, Payload, Slot, SlotType } from "../src/types"

import aBlockAndWhatFollowsIt from "./fixtures/contract/a-block-and-what-follows-it.json"
import aBooleanAttributeDrivenByAValue from "./fixtures/contract/a-boolean-attribute-driven-by-a-value.json"
import aBooleanAttributeInsideABlock from "./fixtures/contract/a-boolean-attribute-inside-a-block.json"
import aCollectionInsideABranch from "./fixtures/contract/a-collection-inside-a-branch.json"
import aConditionalInsideABlock from "./fixtures/contract/a-conditional-inside-a-block.json"
import aConditionalMatchingNothing from "./fixtures/contract/a-conditional-matching-nothing.json"
import aConditionalThatBranches from "./fixtures/contract/a-conditional-that-branches.json"
import aConditionalThatCollapses from "./fixtures/contract/a-conditional-that-collapses.json"
import aHelperBlockBuildingAnElement from "./fixtures/contract/a-helper-block-building-an-element.json"
import aKeyedCollection from "./fixtures/contract/a-keyed-collection.json"
import aKeyedCollectionWithDeclaredItemState from "./fixtures/contract/a-keyed-collection-with-declared-item-state.json"
import aKeyedCollectionWhoseKeysCarryMarkup from "./fixtures/contract/a-keyed-collection-whose-keys-carry-markup.json"
import aValueCarryingMarkupAndQuotes from "./fixtures/contract/a-value-carrying-markup-and-quotes.json"
import anAttributeAndAChild from "./fixtures/contract/an-attribute-and-a-child.json"
import anEmptyCollection from "./fixtures/contract/an-empty-collection.json"
import anInterpolatedAttributeInsideABlock from "./fixtures/contract/an-interpolated-attribute-inside-a-block.json"
import expressionsInOrder from "./fixtures/contract/expressions-in-order.json"
import nothingDynamicAtAll from "./fixtures/contract/nothing-dynamic-at-all.json"
import twoAttributesOnOneElement from "./fixtures/contract/two-attributes-on-one-element.json"

interface SchemaSlot {
  index: number
  type: string
  attribute?: string
  name?: string
}

interface Fixture {
  file: string
  template: string
  rendered: string
  client: string
  values: Payload
  expected: string
  schema: { file: string; identifier: string; version: string; slots: SchemaSlot[] }
  manifest: { file: string; identifier: string; version: string; names: Record<string, number>; parts: Record<string, string[]>; states: unknown }
  dependencies: DependencyMap
}

const FIXTURES: Record<string, Fixture> = {
  "a block and what follows it": aBlockAndWhatFollowsIt as Fixture,
  "a boolean attribute driven by a value": aBooleanAttributeDrivenByAValue as Fixture,
  "a boolean attribute inside a block": aBooleanAttributeInsideABlock as Fixture,
  "a collection inside a branch": aCollectionInsideABranch as Fixture,
  "a conditional inside a block": aConditionalInsideABlock as Fixture,
  "a conditional matching nothing": aConditionalMatchingNothing as Fixture,
  "a conditional that branches": aConditionalThatBranches as Fixture,
  "a conditional that collapses": aConditionalThatCollapses as Fixture,
  "a helper block building an element": aHelperBlockBuildingAnElement as Fixture,
  "a keyed collection": aKeyedCollection as Fixture,
  "a keyed collection with declared item state": aKeyedCollectionWithDeclaredItemState as Fixture,
  "a keyed collection whose keys carry markup": aKeyedCollectionWhoseKeysCarryMarkup as Fixture,
  "an attribute and a child": anAttributeAndAChild as Fixture,
  "an empty collection": anEmptyCollection as Fixture,
  "an interpolated attribute inside a block": anInterpolatedAttributeInsideABlock as Fixture,
  "expressions in order": expressionsInOrder as Fixture,
  "nothing dynamic at all": nothingDynamicAtAll as Fixture,
  "a value carrying markup and quotes": aValueCarryingMarkupAndQuotes as Fixture,
  "two attributes on one element": twoAttributesOnOneElement as Fixture,
}

const PARKED_REASONS = ["branch", "block", "items", "partial-attribute"]

const UNCHANGED = ["nothing dynamic at all", "a boolean attribute inside a block"]

function mount(markup: string, fixture?: Fixture): HTMLElement {
  document.body.innerHTML = ""

  const host = document.createElement("div")
  const manifest = fixture ? container(fixture) : ""

  host.innerHTML = markup + manifest
  document.body.appendChild(host)

  return host
}

function container(fixture: Fixture): string {
  const manifest = fixture.manifest

  return `<template data-herb-manifests>${JSON.stringify({ [`${manifest.identifier}:${manifest.version}`]: manifest })}</template>`
}

function serialized(markup: string): string {
  const holder = document.createElement("div")

  holder.innerHTML = markup

  return holder.innerHTML
}

function found(index: SlotIndex, file: string): Map<number, Slot> {
  const slots = new Map<number, Slot>()

  const visit = (slot: Slot): void => {
    slots.set(slot.index, slot)

    for (const item of slot.items.values()) {
      visitItem(item)
    }
  }

  const visitItem = (item: Item): void => {
    for (const slot of item.slots.values()) {
      visit(slot)
    }
  }

  for (const region of index.regionsFor(file)) {
    for (const slot of region.slots.values()) {
      visit(slot)
    }
  }

  return slots
}

describe("the contract table", () => {
  test("renders every fixture but the static ones differently before and after", () => {
    const same = Object.entries(FIXTURES).filter(([, fixture]) => fixture.rendered === fixture.expected)

    expect(same.map(([label]) => label).sort()).toEqual([...UNCHANGED].sort())
  })
})

describe.each(Object.entries(FIXTURES))("the contract for %s", (_label, fixture) => {
  const schema = new Map(fixture.schema.slots.map((slot) => [slot.index, slot]))

  let index: SlotIndex

  beforeEach(() => {
    index = new SlotIndex()
  })

  test("indexes only slots the compiler recorded, as the type and attribute it recorded", () => {
    index.scan(mount(fixture.rendered, fixture))

    for (const [number, slot] of found(index, fixture.file)) {
      const recorded = schema.get(number)

      expect(recorded, `slot ${number} is on the page but not in the schema`).toBeDefined()
      expect(slot.type).toBe(recorded!.type as SlotType)
      expect(slot.attribute).toBe(recorded!.attribute ?? null)
    }
  })

  test("indexes the client render the same way and takes the parked statics off the page", () => {
    const server = new SlotIndex()
    server.scan(mount(fixture.rendered, fixture))
    const serverSlots = [...found(server, fixture.file).keys()].sort()

    const host = mount(fixture.client, fixture)
    index.scan(host)

    expect([...found(index, fixture.file).keys()].sort()).toEqual(serverSlots)
    expect(host.querySelector("template[data-herb-region]")).toBeNull()
  })

  test("carries the version the payload names", () => {
    index.scan(mount(fixture.rendered, fixture))

    for (const region of index.regionsFor(fixture.file)) {
      expect(region.version).toBe(fixture.values.version)
      expect(region.version).toBe(fixture.schema.version)
    }
  })

  test("becomes the server's next render when the values land on the client render", () => {
    const host = mount(fixture.client, fixture)
    index.scan(host)

    const report = index.apply(fixture.values)

    expect(report.deferred).toEqual([])
    expect(host.innerHTML).toBe(serialized(fixture.expected))
  })

  test("becomes the server's next render on the server render, or defers only what needs parked markup", () => {
    const host = mount(fixture.rendered, fixture)
    index.scan(host)

    const report = index.apply(fixture.values)

    if (report.deferred.length === 0) {
      expect(host.innerHTML).toBe(serialized(fixture.expected))
    } else {
      for (const deferred of report.deferred) {
        expect(PARKED_REASONS).toContain(deferred.reason)
      }
    }
  })

  test("is unchanged by applying the same values a second time", () => {
    const host = mount(fixture.client, fixture)
    index.scan(host)
    index.apply(fixture.values)

    const settled = host.innerHTML

    index.apply(fixture.values)

    expect(host.innerHTML).toBe(settled)
  })

  test("can be indexed again from what it became", () => {
    const host = mount(fixture.client, fixture)
    index.scan(host)
    index.apply(fixture.values)

    const again = new SlotIndex()
    again.scan(host)

    expect([...found(again, fixture.file).keys()].sort()).toEqual([...found(index, fixture.file).keys()].sort())
  })

  test("is described by a dependency map that names only slots the compiler recorded", () => {
    for (const entries of Object.values(fixture.dependencies.state)) {
      for (const entry of entries) {
        expect(entry.file).toBe(fixture.file)
        expect(entry.version).toBe(fixture.schema.version)
        expect(schema.has(entry.index), `the map names slot ${entry.index}, which the schema does not have`).toBe(true)
      }
    }
  })
})

File: test/engine/slots/contract_test.rb
# frozen_string_literal: true

require "json"
require "tmpdir"
require "fileutils"
require "erb/util"
require "active_support/core_ext/string/output_safety"

require_relative "../../test_helper"
require_relative "../../../lib/herb/engine/dynamics_compiler"
require_relative "../../../lib/herb/engine/slot_dependencies"

module Engine
  module Slots
    # Generates the fixtures the client's contract test drives, one per template, so the two
    # implementations are checked against each other on every construct the markers can express
    # and not only on the one template round_trip_test.rb covers.
    class ContractTest < Minitest::Spec
      FIXTURES = File.expand_path("../../../javascript/packages/client/test/fixtures", __dir__)
      CONTRACT = File.join(FIXTURES, "contract")
      MARKERS = File.join(FIXTURES, "markers.json")

      BUFFER = "@output_buffer"

      # How the app compiles a template. ActionView escapes on the way into its buffer rather than
      # in the compiled template, and the values compile escapes the same way in every context so
      # the two halves write the same bytes for the same value.
      ESCAPING = {
        escape: true,
        escapefunc: "::ERB::Util.h",
        attrfunc: nil,
        jsfunc: nil,
        cssfunc: nil,
      }.freeze

      # A view whose helpers capture a block the way ActionView does: the block writes into the
      # buffer, the helper swaps the buffer around the yield and wraps what was written. What a
      # helper returns is markup it produced itself, so it comes back safe and is not escaped again.
      class View
        def initialize(**assigns)
          assigns.each { |name, value| instance_variable_set(:"@#{name}", value) }
        end

        def form_with(**) = "<form>#{capture { |*| yield(Field.new) }}</form>".html_safe
        def tag = TagBuilder.new(self)

        def capture
          outer = @output_buffer
          @output_buffer = +""

          value = yield

          written = @output_buffer
          @output_buffer = outer

          (written.empty? ? value.to_s : written).html_safe
        end

        class Field
          def label = "Name"
        end

        class TagBuilder
          def initialize(view)
            @view = view
          end

          def li(**, &) = "<li>#{@view.capture(&)}</li>".html_safe
        end
      end

      TEMPLATES = {
        "expressions in order" => [
          %(<%= @a %><p><%= @b %></p><%= @c %>),
          { a: "a", b: "b", c: "c" },
          { a: "A", b: "B", c: "C" }
        ],
        "an attribute and a child" => [
          %(<li class="<%= @c %>"><%= @n %></li>),
          { c: "item", n: "Marco" },
          { c: "row", n: "Joe" }
        ],
        "two attributes on one element" => [
          %(<div class="<%= @c %>" id="<%= @i %>"></div>),
          { c: "c", i: "i" },
          { c: "d", i: "j" }
        ],
        "a block and what follows it" => [
          %(<%= form_with(model: 1) do |f| %><%= f.label %><% end %><%= @after %>),
          { after: "A" },
          { after: "B" }
        ],
        "a conditional that branches" => [
          %(<% if @on %><b><%= @x %></b><% else %><%= @y %><% end %><%= @z %>),
          { on: true, x: "x", y: "y", z: "z" },
          { on: false, x: "x", y: "Y", z: "Z" }
        ],
        "a conditional that collapses" => [
          %(<% if @on %><h1><%= @x %></h1><% else %><h1><%= @y %></h1><% end %>),
          { on: false, x: "x", y: "y" },
          { on: true, x: "X", y: "y" }
        ],
        "a conditional matching nothing" => [
          %(<% if @on %><%= @x %><% end %>),
          { on: false, x: "x" },
          { on: true, x: "X" }
        ],
        "a keyed collection" => [
          %(<% @items.each do |r| %><li id="<%= r %>"><%= r %></li><% end %>),
          { items: [1, 2] },
          { items: [2, 3] }
        ],
        "a keyed collection whose keys carry markup" => [
          %(<% @items.each do |r| %><li id="<%= r %>"><%= r %></li><% end %>),
          { items: ["a&b", "c<d"] },
          { items: ["c<d", %(e"f)] }
        ],
        "an empty collection" => [
          %(<% @items.each do |r| %><li id="<%= r %>"><%= r %></li><% end %>),
          { items: [] },
          { items: [1] }
        ],
        "a collection inside a branch" => [
          %(<% if @on %><% @items.each do |r| %><li id="<%= r %>"><%= r %></li><% end %><% end %>),
          { on: true, items: ["a"] },
          { on: true, items: ["a", "b"] }
        ],
        "a helper block building an element" => [
          %(<%= tag.li(id: 1) do %><%= @n %><% end %><%= @after %>),
          { n: "n", after: "A" },
          { n: "N", after: "B" }
        ],
        "a conditional inside a block" => [
          %(<%= form_with(model: 1) do |f| %><% if @on %><%= @x %><% end %><% end %>),
          { on: true, x: "x" },
          { on: true, x: "X" }
        ],
        "a boolean attribute driven by a value" => [
          %(<video muted="<%= @muted %>"></video>),
          { muted: false },
          { muted: true }
        ],
        "a value carrying markup and quotes" => [
          %(<li title="<%= @t %>"><%= @n %></li>),
          { t: %(a "quoted" & <tag>), n: "Tom & <b>Jerry</b>" },
          { t: "plain", n: "Tom & Jerry" }
        ],
        "a boolean attribute inside a block" => [
          %(<%# herb:state (sending: false) %><%= form_with(model: 1) do |f| %><video muted="<%= sending %>"></video><% end %>),
          {},
          {}
        ],
        "an interpolated attribute inside a block" => [
          %(<%= form_with(model: 1) do |f| %><li id="row_<%= @r %>">x</li><% end %>),
          { r: 7 },
          { r: 8 }
        ],
        "a keyed collection with declared item state" => [
          %(<%# herb:state (filter: "all") %><ul><% @items.each do |r| %><%# herb:key r %><%# herb:state (starred: false) %><li id="<%= r %>" hidden="<%= filter == "starred" && starred == false %>"><%= r %></li><% end %></ul>),
          { items: ["a", "b"] },
          { items: ["b", "c"] }
        ],
        "nothing dynamic at all" => [
          %(<p>static</p>),
          {},
          {}
        ],
      }.freeze

      def setup
        @project_path = Dir.mktmpdir("herb_slot_contract")
      end

      def teardown
        FileUtils.rm_rf(@project_path)
      end

      def slug(label)
        label.downcase.gsub(/[^a-z0-9]+/, "-").gsub(/\A-|-\z/, "")
      end

      def file_for(label)
        "app/views/contract/#{slug(label)}.html.erb"
      end

      def write(label, source)
        path = File.join(@project_path, file_for(label))

        FileUtils.mkdir_p(File.dirname(path))
        File.write(path, source)

        path
      end

      def render(label, source, assigns, mode:)
        visitor = Herb::Engine::SlotVisitor.new(mode: mode)
        compiled = Herb::Engine.new(source, visitors: [visitor], filename: file_for(label), project_path: @project_path, bufvar: BUFFER, **ESCAPING).src

        [View.new(**assigns).instance_eval(compiled), visitor]
      end

      def values(label, source, assigns)
        compiled = Herb::Engine::DynamicsCompiler.new(source, filename: file_for(label), project_path: @project_path, bufvar: BUFFER, **ESCAPING).src

        View.new(**assigns).instance_eval(compiled)
      end

      def dependencies(label, source)
        path = write(label, source)

        Herb::Engine::SlotDependencies.new(@project_path).payload(path)
      end

      def fixture(label, source, before, after)
        rendered, visitor = render(label, source, before, mode: :server)
        client, = render(label, source, before, mode: :client)
        expected, = render(label, source, after, mode: :server)

        {
          file: file_for(label),
          template: source,
          rendered: rendered,
          client: client,
          values: values(label, source, after),
          expected: expected,
          schema: visitor.schema,
          dependencies: dependencies(label, source),
        }
      end

      def check(path, generated)
        File.write(path, generated) if ENV["UPDATE_SNAPSHOTS"] || ENV["FORCE_UPDATE_SNAPSHOTS"]

        assert_equal generated, File.read(path), "#{File.basename(path)} no longer describes the compiler's output, re-run with UPDATE_SNAPSHOTS=1"
      end

      TEMPLATES.each do |label, (source, before, after)|
        test "the contract fixture for #{label} is the one the compiler produces" do
          FileUtils.mkdir_p(CONTRACT)

          check(File.join(CONTRACT, "#{slug(label)}.json"), "#{JSON.pretty_generate(fixture(label, source, before, after))}\n")
        end
      end

      test "no contract fixture describes a template this table no longer has" do
        skip "fixtures are being regenerated" if ENV["UPDATE_SNAPSHOTS"] || ENV["FORCE_UPDATE_SNAPSHOTS"]

        expected = TEMPLATES.keys.map { |label| "#{slug(label)}.json" }.sort

        assert_equal expected, Dir.children(CONTRACT).sort
      end

      test "the marker grammar fixture is the one SlotMarkers builds" do
        markers = Herb::Engine::SlotMarkers.new

        grammar = {
          slot_open: { child: markers.slot_open(3, :child), conditional: markers.slot_open(4, :conditional), collection: markers.slot_open(5, :collection) },
          slot_close: markers.slot_close(3),
          element_anchors: markers.element_anchors([[6, :attribute, "class"], [7, :child, nil], [8, :boolean_attribute, "muted"]]),
          branch: markers.branch(4, 1),
          item_open: "#{markers.item_open_prefix(5)}a&b#{markers.item_open_suffix}",
          item_close: markers.item_close(5),
          region_open: "#{markers.region_open_prefix("app/views/posts/_card.html.erb", "aaaaaaaa")}2#{markers.region_open_suffix}",
          region_close: markers.region_close("app/views/posts/_card.html.erb"),
          statics_open: markers.statics_open("app/views/posts/_card.html.erb", "aaaaaaaa"),
          statics_close: markers.statics_close,
        }

        check(MARKERS, "#{JSON.pretty_generate(grammar)}\n")
      end
    end
  end
end
